src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/importer.h"
+
+#include <algorithm>
+#include <cmath>
+#include <fstream>
+#include <libxml/uri.h>
+#include <sstream>
+#include <stdexcept>
+
+#include "libcellml/component.h"
+#include "libcellml/importsource.h"
+#include "libcellml/model.h"
+#include "libcellml/parser.h"
+#include "libcellml/reset.h"
+#include "libcellml/units.h"
+#include "libcellml/variable.h"
+
+#include "namespaces.h"
+#include "utilities.h"
+#include "xmldoc.h"
+#include "xmlutils.h"
+
+namespace libcellml {
+
+/**
+ * @brief The Importer::ImporterImpl struct.
+ *
+ * This struct is the private implementation struct for the Importer class.  Separating
+ * the implementation from the definition allows for greater flexibility when
+ * distributing the code.
+ */
+struct Importer::ImporterImpl
+{
+    Importer *mImporter = nullptr;
+
+    ImportLibrary mLibrary;
+    std::vector<std::string> mLibraryIndex;
+    std::vector<std::pair<std::string, std::string>> mExternals;
+
+    bool resolveImport(const ImportedEntityPtr &importedEntity,
+                       const std::string &destination,
+                       const std::string &typeString,
+                       const std::string &baseFile,
+                       std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool resolveComponentImports(const ComponentEntityPtr &parentComponentEntity,
+                                 const std::string &baseFile,
+                                 std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool doResolveImports(ModelPtr &model, const std::string &baseFile,
+                          std::vector<std::tuple<std::string, std::string, std::string>> &history);
+};
+
+Importer::Importer()
+    : mPimpl(new ImporterImpl())
+{
+    mPimpl->mImporter = this;
+}
```

Comment:
I know that this kind of construct is done elsewhere, but... not simply pass `this` as an argument to `ImporterImpl`?

Reply:
I'm just following the lead ...

Reply:
Doesn't make it right.

Reply:
You're welcome to make an issue for it so we can fix it everywhere.